src/neoxp/Commands/ContractCommand.Download.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.IO;
+using System.Net.Http;
+using System.Threading.Tasks;
+using McMaster.Extensions.CommandLineUtils;
+using Neo;
+using Neo.IO.Json;
+using Neo.Network.RPC;
+
+namespace NeoExpress.Commands
+{
+    partial class ContractCommand
+    {
+        [Command(Name = "download", Description = "Download contract with storage from remote chain into local chain")]
+        internal class Download
+        {
+            readonly ExpressChainManagerFactory chainManagerFactory;
+            
+            public Download(ExpressChainManagerFactory chainManagerFactory)
+            {
+                this.chainManagerFactory = chainManagerFactory;
+            }
+            
+            [Argument(0, Description = "Contract invocation hash")]
+            [Required]
+            internal string Contract { get; init; } = string.Empty;
+            
+            [Argument(1, Description = "URL of Neo JSON-RPC Node\nSpecify MainNet (default), TestNet or JSON-RPC URL")]
+            internal string RpcUri { get; } = string.Empty;
+            
+            [Option(Description = "Path to neo-express data file")]
+            internal string Input { get; init; } = string.Empty;
+
+            internal async Task ExecuteAsync(TextWriter writer)
+            {
+                var (chainManager, _) = chainManagerFactory.LoadChain(Input);
+                var expressNode = chainManager.GetExpressNode();
+
+                if (!UInt160.TryParse(Contract, out var contractHash))
+                    throw new ArgumentException($"Invalid contract hash: \"{Contract}\"");
+
+                if (!TransactionExecutor.TryParseRpcUri(RpcUri, out var uri))
+                    throw new ArgumentException($"Invalid RpcUri value \"{RpcUri}\"");
+
+                using var rpcClient = new RpcClient(uri);
+                var stateAPI = new StateAPI(rpcClient);
+
+                var stateHeight = await stateAPI.GetStateHeightAsync();
+                if (stateHeight.localRootIndex is null)
+                    throw new Exception("Null \"localRootIndex\" in state height response");
+                var stateRoot = await stateAPI.GetStateRootAsync(stateHeight.localRootIndex.Value);
+                var states = await stateAPI.FindStatesAsync(stateRoot.RootHash, contractHash, new byte[0]);
```

Comment:
Since you're going to have to update this code to handle `Truncated`...

I'm thinking it would make more sense *NOT* to decode the key/value pairs here. `FindStatesAsync` method decodes the storage pairs from base64 -> byte arrays and then `OnlineNode` turns around and re-encodes them in base64. Maybe it makes more sense to roll our own `FindStatesAsync` implementation that handles both `Truncated` and skips decoding storage items.  `IExpressNode.PersistContractAsync` signature would need to be updated and we could do Base64 -> byte[] decoding in `PersistContract`

Normally, I'm not a fan of leaking implementation details into an interface definition, but in this case, efficiency concerns might trump cleaner design.

Reply:
I pointed out above that the user should be able to specify block height for downloading contract + state, both for normal direct use and also for `batch` mode use. If the code knows the block height to download, we can move all the contract state download code into the `ExpressOracle.PersistContract` method. We wouldn't need to pass all the downloaded state to `ExpressPersistContract`, just the block height

Reply:
I was not happy with the decode, re-encode either but submitted as is to see what you found acceptable. 

As for moving everything to `ExpressOracle.PersistContract` it would end up requiring `RpcUri` and `block height` to be pushed down. If that's OK with you I'll make it happen. 

How about making `block height` optional and have it default to `-1` meaning get the latest state?

Reply:
I'm fine with pushing `RpcUri` and `Height` down into `PersistContract`, but I think the calling code needs to pass a valid block height. For direct mode, if the user doesn't pass a specific height, the command execution code can retrieve the current height to pass to `PersistContract`. For batch mode, the batch height is required.